File: core/internal/roster/roster.go
// Package roster implements the device roster: the abstraction that unifies a
// user's devices under the two-layer identity model (data-sync-decisions §4.2
// P3). Each device has its own device DID (its libp2p transport key, so
// peerID ≡ deviceDID) while the user DID is the account visible to the network.
// A roster lists the user's device DIDs and is signed by the USER key, so any
// holder of the user DID can verify which devices belong to that user.
//
// Wire-compatible with the TypeScript port (ts/linkself/src/roster.ts): the
// canonical bytes signed are the compact JSON {v,userDID,devices:[{deviceDID,
// label}]} with devices sorted by deviceDID, and Marshal produces
// {userDID,devices,sig} with sig base64-encoded.
package roster

import (
	"encoding/json"
	"sort"

	"github.com/SeijiShii/link-self/core/internal/did"
)

// DeviceEntry is one device in a user's roster.
type DeviceEntry struct {
	// DeviceDID is the device's own did:key (its libp2p transport identity).
	DeviceDID string `json:"deviceDID"`
	// Label is a human-readable name (e.g. "PC"); empty string if unset.
	Label string `json:"label"`
}

// SignedRoster is a user's device list, signed by the user key.
type SignedRoster struct {
	// UserDID is the account DID — the signer, visible to the network.
	UserDID string `json:"userDID"`
	// Devices are all device DIDs belonging to this user.
	Devices []DeviceEntry `json:"devices"`
	// Sig is the Ed25519 signature by the user key over the canonical encoding.
	// json.Marshal encodes it as base64 (matching the TS wire format).
	Sig []byte `json:"sig"`
}

// canonicalBytes is the deterministic encoding signed by the user key. Devices
// are sorted by deviceDID so the same roster always yields the same bytes.
func canonicalBytes(userDID string, devices []DeviceEntry) ([]byte, error) {
	sorted := append([]DeviceEntry(nil), devices...)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].DeviceDID < sorted[j].DeviceDID
	})
	return json.Marshal(struct {
		V       int           `json:"v"`
		UserDID string        `json:"userDID"`
		Devices []DeviceEntry `json:"devices"`
	}{V: 1, UserDID: userDID, Devices: sorted})
}

// Build builds and signs a roster with the user identity. Devices are
// de-duplicated by deviceDID (last wins).
func Build(user *did.Identity, devices []DeviceEntry) (*SignedRoster, error) {
	deduped := dedupeByDeviceDID(devices)
	b, err := canonicalBytes(user.DID, deduped)
	if err != nil {
		return nil, err
	}
	sig, err := user.PrivKey.Sign(b)
	if err != nil {
		return nil, err
	}
	return &SignedRoster{UserDID: user.DID, Devices: deduped, Sig: sig}, nil
}

// Verify checks the roster signature against its own UserDID's public key.
func Verify(r *SignedRoster) bool {
	pub, err := did.ParseToPubKey(r.UserDID)
	if err != nil {
		return false
	}
	b, err := canonicalBytes(r.UserDID, r.Devices)
	if err != nil {
		return false
	}
	ok, err := pub.Verify(b, r.Sig)
	return err == nil && ok
}

// HasDevice reports whether a device DID is a member of the roster.
func HasDevice(r *SignedRoster, deviceDID string) bool {
	for _, d := range r.Devices {
		if d.DeviceDID == deviceDID {
			return true
		}
	}
	return false
}

// DeviceDIDs returns the device DIDs in the roster.
func DeviceDIDs(r *SignedRoster) []string {
	out := make([]string, len(r.Devices))
	for i, d := range r.Devices {
		out[i] = d.DeviceDID
	}
	return out
}

// WithDevice returns a new signed roster with entry added or its label updated.
func WithDevice(user *did.Identity, devices []DeviceEntry, entry DeviceEntry) (*SignedRoster, error) {
	next := make([]DeviceEntry, 0, len(devices)+1)
	for _, d := range devices {
		if d.DeviceDID != entry.DeviceDID {
			next = append(next, d)
		}
	}
	next = append(next, entry)
	return Build(user, next)
}

// WithoutDevice returns a new signed roster with the given device removed.
func WithoutDevice(user *did.Identity, devices []DeviceEntry, deviceDID string) (*SignedRoster, error) {
	next := make([]DeviceEntry, 0, len(devices))
	for _, d := range devices {
		if d.DeviceDID != deviceDID {
			next = append(next, d)
		}
	}
	return Build(user, next)
}

// Marshal serializes a roster to JSON (sig base64), for storage or transfer.
func Marshal(r *SignedRoster) ([]byte, error) {
	return json.Marshal(r)
}

// Unmarshal parses a roster from JSON produced by Marshal.
func Unmarshal(data []byte) (*SignedRoster, error) {
	var r SignedRoster
	if err := json.Unmarshal(data, &r); err != nil {
		return nil, err
	}
	return &r, nil
}

func dedupeByDeviceDID(devices []DeviceEntry) []DeviceEntry {
	seen := make(map[string]int, len(devices)) // deviceDID -> index in out
	out := make([]DeviceEntry, 0, len(devices))
	for _, d := range devices {
		if i, ok := seen[d.DeviceDID]; ok {
			out[i] = d // last wins
			continue
		}
		seen[d.DeviceDID] = len(out)
		out = append(out, d)
	}
	return out
}

File: ts/linkself/src/roster.ts
/**
 * Device roster: the abstraction that unifies a user's devices.
 *
 * Two-layer identity model (data-sync-decisions §4.2 P3): each device has its
 * own device DID (its libp2p transport key, so peerId ≡ deviceDID), while the
 * user DID is the account visible to the network. A roster binds them: it lists
 * the user's device DIDs and is signed by the USER key, so anyone holding the
 * user DID (its public key) can verify which devices legitimately belong to
 * that user.
 *
 * This lets device-to-device trust skip a challenge-response: Noise already
 * authenticates the transport key (= peer's device DID), and roster membership
 * (verified against the user key) proves that device DID is a sibling of the
 * same user. Routing/app layers operate on the user DID; the transport layer
 * on device DIDs; the roster bridges the two.
 */
import { parseToPublicKey, type Identity } from "./did.js";

/** One device in a user's roster. */
export interface DeviceEntry {
  /** The device's own did:key (its libp2p transport identity). */
  deviceDID: string;
  /** Human-readable label (e.g. "PC", "Phone"). Empty string if unset. */
  label: string;
}

/** A user's device list, signed by the user key. */
export interface SignedRoster {
  /** The user (account) DID — the signer, visible to the network. */
  userDID: string;
  /** All device DIDs belonging to this user. */
  devices: DeviceEntry[];
  /** Ed25519 signature by the user key over the canonical encoding. */
  sig: Uint8Array;
}

/**
 * Deterministic byte encoding signed by the user key. Devices are sorted by
 * deviceDID so the same roster always yields the same bytes regardless of
 * insertion order.
 */
function canonicalBytes(userDID: string, devices: DeviceEntry[]): Uint8Array {
  const sorted = [...devices].sort((a, b) =>
    a.deviceDID < b.deviceDID ? -1 : a.deviceDID > b.deviceDID ? 1 : 0,
  );
  const obj = {
    v: 1,
    userDID,
    devices: sorted.map((d) => ({ deviceDID: d.deviceDID, label: d.label })),
  };
  return new TextEncoder().encode(JSON.stringify(obj));
}

/** Build and sign a roster with the user identity. Devices are de-duplicated by deviceDID. */
export async function buildRoster(
  userIdentity: Identity,
  devices: DeviceEntry[],
): Promise<SignedRoster> {
  const deduped = dedupeByDeviceDID(devices);
  const sig = await userIdentity.privateKey.sign(
    canonicalBytes(userIdentity.did, deduped),
  );
  return { userDID: userIdentity.did, devices: deduped, sig };
}

/** Verify the roster signature against its own userDID's public key. */
export async function verifyRoster(roster: SignedRoster): Promise<boolean> {
  try {
    const pub = parseToPublicKey(roster.userDID);
    return await pub.verify(
      canonicalBytes(roster.userDID, roster.devices),
      roster.sig,
    );
  } catch {
    return false;
  }
}

/** The device DIDs in the roster. */
export function rosterDeviceDIDs(roster: SignedRoster): string[] {
  return roster.devices.map((d) => d.deviceDID);
}

/** Whether a device DID is a member of the roster. */
export function rosterHasDevice(
  roster: SignedRoster,
  deviceDID: string,
): boolean {
  return roster.devices.some((d) => d.deviceDID === deviceDID);
}

/**
 * Return a new signed roster with `entry` added or its label updated. Requires
 * the user identity (every device holds the user key after pairing, so any
 * device can amend and re-sign the roster).
 */
export async function withDevice(
  userIdentity: Identity,
  devices: DeviceEntry[],
  entry: DeviceEntry,
): Promise<SignedRoster> {
  const next = devices.filter((d) => d.deviceDID !== entry.deviceDID);
  next.push(entry);
  return buildRoster(userIdentity, next);
}

/** Return a new signed roster with the given device removed (roster revocation). */
export async function withoutDevice(
  userIdentity: Identity,
  devices: DeviceEntry[],
  deviceDID: string,
): Promise<SignedRoster> {
  return buildRoster(
    userIdentity,
    devices.filter((d) => d.deviceDID !== deviceDID),
  );
}

function dedupeByDeviceDID(devices: DeviceEntry[]): DeviceEntry[] {
  const byDID = new Map<string, DeviceEntry>();
  for (const d of devices) byDID.set(d.deviceDID, d); // last wins
  return [...byDID.values()];
}

/* ---- Wire encoding (for storage in MyDB / transfer) ---- */

/** Serialize a roster to JSON bytes (sig as base64). */
export function marshalRoster(roster: SignedRoster): Uint8Array {
  const obj = {
    userDID: roster.userDID,
    devices: roster.devices,
    sig: bytesToBase64(roster.sig),
  };
  return new TextEncoder().encode(JSON.stringify(obj));
}

/** Parse a roster from JSON bytes produced by marshalRoster. */
export function unmarshalRoster(data: Uint8Array): SignedRoster {
  const p = JSON.parse(new TextDecoder().decode(data)) as {
    userDID: string;
    devices: DeviceEntry[];
    sig: string;
  };
  return {
    userDID: p.userDID,
    devices: p.devices ?? [],
    sig: base64ToBytes(p.sig),
  };
}

function bytesToBase64(bytes: Uint8Array): string {
  let s = "";
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s);
}

function base64ToBytes(b64: string): Uint8Array {
  const s = atob(b64);
  const out = new Uint8Array(s.length);
  for (let i = 0; i < s.length; i++) out[i] = s.charCodeAt(i);
  return out;
}
